### PR TITLE
Onnx 1.8 support

### PR DESCRIPTION
**Pull request by scxiao**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -80,6 +78,9 @@ RUN cget -p $PREFIX install ccache@v4.1
 # Install newer cmake for onnx runtime
 RUN cget -p /opt/cmake install kitware/cmake@v3.13.0
 
+# Install onnx 1.8.1
+RUN pip3 install onnx numpy typing pytest
```
Comment (pfultz2) [inline Dockerfile:82]: Isn't this already installed in the `install_prereqs.sh` script?

File: src/include/migraphx/op/logsoftmax.hpp [hunk 2]
```diff
@@ -12,7 +12,7 @@ namespace op {
 
 struct logsoftmax
 {
-    int64_t axis = 1;
+    int64_t axis = -1;
```
Comment (pfultz2) [inline src/include/migraphx/op/logsoftmax.hpp:15]: Why are we changing the default axis?

File: src/onnx/parse_squeeze.cpp [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        return {{"Squeeze", "squeeze"}, {"Unsqueeze", "unsqueeze"}};
+    }
+
+    bool needs_contiguous(const std::string& op_name) const
+    {
+        return contains({"squeeze", "unsqueeze"}, op_name);
+    }
+
+    void assign_axes(const operation& op, const std::vector<int64_t>& axes) const
```
Comment (pfultz2) [inline src/onnx/parse_squeeze.cpp:23]: This function looks like a nop. What is it supposed to be doing?

File: src/onnx/parse_squeeze.cpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+namespace onnx {
+
+struct parse_squeeze : op_parser<parse_squeeze>
+{
+    std::vector<op_desc> operators() const
+    {
+        return {{"Squeeze", "squeeze"}, {"Unsqueeze", "unsqueeze"}};
+    }
+
+    bool needs_contiguous(const std::string& op_name) const
```
Comment (pfultz2) [inline src/onnx/parse_squeeze.cpp:18]: All the operators listed need contiguous, so this function doesn't seem necessary.

File: Dockerfile (hunk 1 above)
Comment (pfultz2) [reply Dockerfile:82]: Also, the pinned versions should not be dropped.

Comment (scxiao) [reply Dockerfile:82]: Yes, you are right. I removed this line and the line for 1.7.0 in the Dockerfile. 

Actually, I am wondering whether we can install two python versions (like 3.6 and 3.7), and install different onnx versions on them?

File: src/include/migraphx/op/logsoftmax.hpp (hunk 2 above)
Comment (scxiao) [reply src/include/migraphx/op/logsoftmax.hpp:15]: The onnx 1.8.1 changed the default value from 1 to -1.

File: src/onnx/parse_squeeze.cpp (hunk 3 above)
Comment (scxiao) [reply src/onnx/parse_squeeze.cpp:23]: I forgot a return value. Fixed.

File: src/onnx/parse_squeeze.cpp (hunk 4 above)
Comment (scxiao) [reply src/onnx/parse_squeeze.cpp:18]: fixed.

File: src/include/migraphx/op/logsoftmax.hpp (hunk 2 above)
Comment (pfultz2) [reply src/include/migraphx/op/logsoftmax.hpp:15]: Then this should be changed in the onnx parser.

File: Dockerfile (hunk 1 above)
Comment (pfultz2) [reply Dockerfile:82]: Is the reason to install two different python version so we can have two different onnx versions installed? As we can create two different venvs to install onnx. I dont think we should do that in this PR, but we could look into cmake automatically setting up the venvs so the the tests will know where to find them.

File: src/onnx/parse_quantizelinear.cpp [hunk 5]
```diff
@@ -72,6 +72,21 @@ struct parse_quantizelinear : op_parser<parse_quantizelinear>
...
+        if(min_arg->get_shape() != s)
+        {
+            min_arg = info.add_instruction(make_op("multibroadcast", {{"output_lens", out_lens}}),
+                                           min_arg);
+        }
+        if(max_arg->get_shape() != s)
+        {
+            max_arg = info.add_instruction(make_op("multibroadcast", {{"output_lens", out_lens}}),
+                                           max_arg);
+        }
```
Comment (pfultz2) [inline src/onnx/parse_quantizelinear.cpp:88]: This can use `add_broadcastable_binary_op` instead.

File: src/onnx/parse_reduce_op.cpp [hunk 6]
```diff
@@ -15,15 +16,42 @@ instruction_ref parse_reduce_oper(const std::string& op_name,
...
+        arg_axes.visit([&](auto s) { axes.assign(s.begin(), s.end()); });
+    }
+    else if(contains(info.attributes, "axes"))
     {
         axes.clear();
         auto&& attr_axes = info.attributes["axes"].ints();
         axes             = std::vector<int64_t>(attr_axes.begin(), attr_axes.end());
     }
 
+    int noop_with_empty_axes = 0;
```
Comment (pfultz2) [inline src/onnx/parse_reduce_op.cpp:34]: Shouldn't this be a `bool`?

File: src/onnx/parse_squeeze.cpp [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        auto v = op.to_value();
+        for(auto&& x : v)
+        {
+            if(x.get_key() == "axes")
+            {
+                x = axes;
+                break;
+            }
+        }
```
Comment (pfultz2) [inline src/onnx/parse_squeeze.cpp:28]: Why not write this as `v["axes"] = axes`?

File: Dockerfile (hunk 1 above)
Comment (scxiao) [reply Dockerfile:82]: yes, we want two different onnx versions installed, so we can test both onnx versions.
OK, we can push later to add that.

File: src/include/migraphx/op/logsoftmax.hpp (hunk 2 above)
Comment (scxiao) [reply src/include/migraphx/op/logsoftmax.hpp:15]: So, we will set different default values for different onnx versions, right?

File: src/onnx/parse_quantizelinear.cpp (hunk 5 above)
Comment (scxiao) [reply src/onnx/parse_quantizelinear.cpp:88]: The "clip" here is not a binary, we add another function for the broadcasting?

File: src/onnx/parse_reduce_op.cpp (hunk 6 above)
Comment (scxiao) [reply src/onnx/parse_reduce_op.cpp:34]: The ONNX operator uses an int type, but I can change it to bool here.

File: src/onnx/parse_squeeze.cpp (hunk 7 above)
Comment (scxiao) [reply src/onnx/parse_squeeze.cpp:28]: changed

File: src/include/migraphx/op/logsoftmax.hpp (hunk 2 above)
Comment (pfultz2) [reply src/include/migraphx/op/logsoftmax.hpp:15]: > So, we will set different default values for different onnx versions, right?

If we want backwards compatibility to 1.7 then we will probably need to do this. More importantly, onnx may change default values but that should not necessarily change our default values. Rather, the onnx parser can decide a default value that is different than ours.

Is the onnx version saved in the onnx file?

File: tools/install_prereqs.sh [hunk 8]
```diff
@@ -6,7 +6,7 @@
 apt update && apt install -y python3-pip rocm-cmake rocblas miopen-hip openmp-extras
 
 # install onnx package for unit tests
-pip3 install onnx==1.7.0 numpy==1.18.5 typing==3.7.4 pytest==6.0.1
+pip3 install onnx==1.8.1 numpy==1.18.5 typing==3.7.4 pytest==6.0.1 packaging
```
Comment (pfultz2) [inline tools/install_prereqs.sh:9]: `packaging` version should be pinned.

File: src/onnx/onnx_parser.cpp [hunk 9]
```diff
@@ -29,6 +29,12 @@ static onnx_parser::attribute_map get_attributes(const onnx::NodeProto& node)
 static literal
 create_literal(shape::type_t shape_type, const std::vector<size_t>& dims, const char* data)
 {
+    // empty input
+    auto elem_num =
+        std::accumulate(dims.begin(), dims.end(), std::size_t(1), std::multiplies<std::size_t>());
+    if(elem_num == 0)
+        return {};
```
Comment (pfultz2) [inline src/onnx/onnx_parser.cpp:36]: There should be a test for this.

**Comment by scxiao** [comment]
@pfultz2 @kahmed10 @cagery Can you review this PR?

File: Dockerfile [hunk 10]
```diff
@@ -73,6 +71,8 @@ ADD requirements.txt /requirements.txt
 COPY ./tools/install_prereqs.sh /
 RUN /install_prereqs.sh /usr/local / && rm /install_prereqs.sh
 
+RUN pip3 install onnx==1.8.1 numpy==1.18.5 typing==3.7.4 pytest==6.0.1 packaging
```
[the remark this answers is not in the document]
Comment (kahmed10) [reply Dockerfile]: we should probably refactor our Dockerfile as a whole to make sure there aren't duplicates in install_prereqs.sh

File: src/onnx/onnx_parser.cpp [hunk 11]
```diff
@@ -238,6 +257,26 @@ void onnx_parser::parse_from(const void* data, std::size_t size)
...
 
+void onnx_parser::parse_opset_version(const onnx::ModelProto& model)
+{
+    const auto& opset_import = model.opset_import();
+    int64_t version          = -1;
+    for(const auto& opset : opset_import)
+    {
+        if(opset.has_version())
+        {
+            version = (version < opset.version()) ? opset.version() : version;
```
Comment (kahmed10) [inline src/onnx/onnx_parser.cpp:268]: can we rewrite this as `version = std::max(version, opset.version())`?

File: src/onnx/onnx_parser.cpp [hunk 12]
```diff
@@ -238,6 +257,26 @@ void onnx_parser::parse_from(const void* data, std::size_t size)
     }
 }
 
+void onnx_parser::parse_opset_version(const onnx::ModelProto& model)
```
Comment (pfultz2) [inline src/onnx/onnx_parser.cpp:260]: Can we make this a static function that returns the opset version: `std::size_t get_opset_version(const onnx::ModelProto& model)`? And then we can set the version with `opset_version = get_opset_version(model)`.

File: tools/install_prereqs.sh (hunk 8 above)
Comment (scxiao) [reply tools/install_prereqs.sh]: pinned

File: Dockerfile (hunk 10 above)
[the remark this answers is not in the document]
Comment (scxiao) [reply Dockerfile]: just checked the docker file, there should be no duplicates anymore.

File: src/onnx/onnx_parser.cpp (hunk 9 above)
Comment (scxiao) [reply src/onnx/onnx_parser.cpp:37]: added one

File: src/onnx/onnx_parser.cpp (hunk 12 above)
Comment (scxiao) [reply src/onnx/onnx_parser.cpp]: changed accordly

File: src/onnx/onnx_parser.cpp (hunk 11 above)
Comment (scxiao) [reply src/onnx/onnx_parser.cpp]: changed

**Review by pfultz2 — APPROVED** [review]

**Review by kahmed10 — APPROVED** [review]

**Comment by scxiao** [comment]
@cagery Can you review this PR when you get a chance? We want to get it in today, thanks!

**Review by cagery — APPROVED** [review]

**Pull request merged**